packages/avatar/contracts/nft-collection/NFTCollection.sol

```diff
@@ -805,16 +805,16 @@ INFTCollection
             SafeERC20.safeTransferFrom($.allowedToExecuteMint, wallet, $.mintTreasury, _price);
         }
         uint256 _totalSupply = $.totalSupply;
+        waveData.waveOwnerToClaimedCounts[wallet] += amount;
+        waveData.waveTotalMinted += amount;
+        $.totalSupply += amount;
+        $.mintedCount[wallet] += amount;
         for (uint256 i; i < amount; i++) {
+            emit WaveMint(_totalSupply + i + 1, wallet, waveIndex);
             // @dev _safeMint already checks the destination _wallet
             // @dev start with tokenId = 1
             _safeMint(wallet, _totalSupply + i + 1);
-            emit WaveMint(_totalSupply + i + 1, wallet, waveIndex);
         }
```

Comment:
I don't think its right to emit mint event before mint function, please move it under _safeMint.

Reply:
Emit is considered a side effect that was the reason to have two loops.
In any case we cannot follow the pattern completely, so.